### ISSUE TITLE
Compacting Incremental Snapshot Files

### ISSUE THREAD

**Issue by afritzler** [gardener/etcd-druid#88]
Question: How are you guys dealing with the incremental backup files when restoring a cluster? I am asking, because in Kubify we expect one full snapshot file to trigger a restore operation. What is the best way to compact all the incremental backup files into one? Do you have already something handy?

**Comment by shreyas-s-rao** [comment]
@majst01 We require incremental snapshots in a general sense to avoid frequent large full snapshots which usually create a network costs. If you do want to avoid delta snapshots, you can simply configure `chart/etcd-backup-restore/values.yaml` by setting the `backup.deltaSnapshotPeriod` value to anything less than `1s` to completely disable delta snapshots, and you can also configure the `backup.schedule` value to set the full snapshot schedule to a higher frequency.

Regarding backup compression, we have opened https://github.com/gardener/etcd-backup-restore/issues/255 and it's on the project roadmap.

**Comment by majst01** [comment]
Thanks @shreyas-s-rao for the hints.

We tried ourselves to set the backup to do full backups only, but from our experience setting `backup.schedule` had no effect. 
Maybe @Gerrit91 can give you more information on that.
We instead modified the generated `etcd` resource manually which is not the way to go.

**Comment by amshuman-kr** [comment]
> We tried ourselves to set the backup to do full backups only, but from our experience setting backup.schedule had no effect.

@majst01 I just tried the following and delta snapshots were disabled with only fullsnapshots enabled.
```yaml
    backup:
        ...
        fullSnapshotSchedule: "* * * * *"
        ...
        deltaSnapshotPeriod: 0s
        ...
```

```
time="2020-10-01T07:09:19Z" level=info msg="Taking scheduled snapshot for time: 2020-10-01 07:09:19.1179478 +0000 UTC" actor=snapshotter
{"level":"warn","ts":"2020-10-01T07:09:19.126Z","caller":"clientv3/retry_interceptor.go:116","msg":"retry stream intercept"}
time="2020-10-01T07:09:19Z" level=info msg="Successfully opened snapshot reader on etcd" actor=snapshotter
time="2020-10-01T07:09:19Z" level=info msg="Total time to save snapshot: 0.004196 seconds."
time="2020-10-01T07:09:19Z" level=info msg="Successfully saved full snapshot at: Backup-1601536159/Full-00000000-00000001-1601536159" actor=snapshotter
time="2020-10-01T07:09:19Z" level=info msg="Will take next full snapshot at time: 2020-10-01 07:10:00 +0000 UTC" actor=snapshotter
time="2020-10-01T07:09:19Z" level=info msg="Setting status to : 200" actor=backup-restore-server
time="2020-10-01T07:09:19Z" level=info msg="Starting snapshotter..." actor=backup-restore-server
time="2020-10-01T07:09:19Z" level=info msg="Taking scheduled snapshot for time: 2020-10-01 07:09:19.132893 +0000 UTC" actor=snapshotter
time="2020-10-01T07:09:19Z" level=info msg="There are no updates since last snapshot, skipping full snapshot." actor=snapshotter
time="2020-10-01T07:09:19Z" level=info msg="Stopping full snapshot..." actor=snapshotter
time="2020-10-01T07:09:19Z" level=info msg="Resetting full snapshot to run after 40.8607142s" actor=snapshotter
```

**Comment by amshuman-kr** [comment]
> I think there is no real benefit of doing incrementals at all, when looking at our environment, full backup is ~100MB

It really depends on the workload pattern in the cluster. In most clusters the delta snapshots are much smaller than 100M but in the seed and garden control-plane, it is quite high. As a consequence, the full snapshot size is also quite high (2-3G).

We have an issue to compress/decompress snapshots (full as well as delta) https://github.com/gardener/etcd-backup-restore/issues/255.

We also have this issue to compact the delta snapshots in the background.

In the special case, the full snapshots are small and comparable to the delta snapshot, it might still make sense to do only full snapshots as you mentioned. Would it make sense to make this configurable in gardener?

**Comment by majst01** [comment]
@amshuman-kr in which resource did you set these setting ?

```
    backup:
        ...
        fullSnapshotSchedule: "* * * * *"
        ...
        deltaSnapshotPeriod: 0s
        ...
```

**Comment by Gerrit91** [comment]
The `etcd` resource is deployed by Gardener and I think only the extension-providers modify the values via webhooks. The gardener-extension-provider-gcp has a `Schedule` field in the [deployment config](https://github.com/gardener/gardener-extension-provider-gcp/blob/v1.11.1/pkg/apis/config/v1alpha1/types.go#L65), but as far as I can see from the code this config flag is not used.

**Comment by amshuman-kr** [comment]
> @amshuman-kr in which resource did you set these setting ?

The `Etcd` resource if `etcd-druid`.

**Comment by amshuman-kr** [comment]
This issue was partially addressed in https://github.com/gardener/etcd-backup-restore/pull/301. The functionality will be complete once #191 is completed.

**Comment by shreyas-s-rao** [comment]
This issue is now fully addressed with https://github.com/gardener/etcd-druid/pull/197. Hence closing it.
The snapshot compaction feature will be available in etcd-druid `v0.6.0` release shortly.
